feat: the entirety of the movies backend API

Kleamens opened the pull request:
### Changes 

The files which were previously deleted from the main branch are copied from code-review branch which contained them before their deletion from main branch 

-- or -- 

+Added endpoints and services  for CRUD operations on directors, movies 

+ Added endpoints for displaying top 5 movies by revenue based on a given genre filter

+Created .csv parsing script for the dataset of movies and directors

## What
The files which were originally in main branch

+ Data layer descriptions of movies and directors

+ Services which interact with movies and directors repository

+ REST Controller for individual endpoints

+ Parsing script for .csv containing information on different movies and their directors

## Why?
There was a need to temporary remove the files from the main branch, now the deleted files are required again

+ There a need for being able to fill the database with the information contained in the .csv file

+ There was a requirement for performing CRUD operation via endpoints

## How?
Pull request which will add the previously deleted files

Kleamens committed:
The entire Movies backend

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,135 @@
+    public dataDownloader(MovieService movieService,DirectorService directorService){
+        this.directorService = directorService;
+        this.movieService = movieService;
+    }
```
:bug: Mate `AllArgsConstructor` anotaci, neni potreba vytvaret konstruktor.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,135 @@
+
+
+@Service
+@NoArgsConstructor(force = true)
```
:bug: Zbytecna anotace. Nechcete prece vytvaret instanci bez zavislosti. Navic si to odporuje s tym, ze ty fieldy mate `final` (directorService a movieService)

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,135 @@
+
+
+    public void downloadData(){
+        String filename ="C:\\Users\\Person\\Desktop\\ea_git\\ea_ls2023-2024_xzirchuk\\moviesAPI\\src\\main\\java\\cz\\mendelu\\ea\\xzirchuk\\project\\moviesAPI\\domain\\data\\imdb_top_1000.csv";
```
🐛 To csv presunout do `resources` slozky a nacitat to cez `*.class.getResource()`

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,135 @@
+                for (int i = 1; i<nextRecord.length; i++) {
+                    params.add(nextRecord[i]);
+                }
+                System.out.println(params);
```
:bug: Namisto `System.out` pouzit nejaku logovaci knihovny. (napríklad. Slf4j)

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,135 @@
+                insertMovie.setGenre(params.get(4)
+                        .replace(" ",""));
+                insertMovie.setImdbRating(Float.parseFloat(
+                        params.get(5)
```
:bug: Vytvořit konstanty s popisnym nazvem. `get(5)` -> `get(RATING)`

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,135 @@
+                System.out.println();
+            }
+        }
+        catch (Exception e) {
```
:bug: Odchytavat konkretnejsi vyjimku nebo vic ruznych vyjimek a na kazdou patricne reagovat.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/director/Director.java:
```diff
@@ -0,0 +1,42 @@
+
+    @OneToMany(mappedBy = "director",cascade = CascadeType.REMOVE)
+    @ToString.Exclude
+    @JsonIgnore
```
:bug: Tohle tady nema byt.

jankoscak commented on moviesAPI/src/main/resources/application.properties:
```diff
@@ -0,0 +1,14 @@
+
+server.port=8090
```
:bug: Proc je potreba jinej nez defaultni port?

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/utils/exceptions/AlreadyExistsException.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.CONFLICT)
```
🐛  Namisto tyhle anotace je potreba napsat `ExceptionHandler`, kterej bude spracovavat vyjimky a bude vracet nejakou `ErrorResponse` nejenom se spravnym status code, ale i s nejakou zpravou co bylo spatne.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/movie/MovieResponse.java:
```diff
@@ -0,0 +1,77 @@
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import cz.mendelu.ea.xzirchuk.project.moviesAPI.domain.director.Director;
+import jakarta.persistence.*;
```
🐛 Neni potreba mit tady `jakarta.persistence`

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/movie/MovieContoller.java:
```diff
@@ -0,0 +1,270 @@
+            @RequestParam(required = false) List<String> genre,
+            @RequestParam(required = false) String year
+            ){
+        var movies = movieService.getAllMovies();
+       String string_genres = genre.stream()
+                .map(n -> String.valueOf(n))
+                .collect(Collectors.joining(",", "", ""));
+
+        if(movieService.doesGenreContainInvalid(string_genres)){
+            throw new BadInputException();
+        }
+//        filtering by req params
+        var filtered_movies= movieService.filterMoviesByParamters(
+                director_name,
+                year!=null?LocalDate.parse(year):null,
+                genre,
+                movies
+        );
+        if (filtered_movies.isEmpty()){throw new NotFoundException();};
+//sorting by revenue
+        movieService.sortMoviesByRevenue(filtered_movies);
+//getting first n results
+        try{
+            filtered_movies=filtered_movies.subList(0,Integer.parseInt(top_n));
+        }catch (IndexOutOfBoundsException e){
+            System.out.println("### LIMIT TOO HIGH, RETURNING ALL OF THE AVAILABLE ITEMS ");
+        }catch (NumberFormatException e){
+            throw  new BadInputException();
+        }
+        return ArrayResponse.of(filtered_movies,MovieResponse::new);
```
:bug: Kopec logiky, ktera by mela byt v service a ne v controlleri.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/movie/Movie.java:
```diff
@@ -0,0 +1,79 @@
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.NotNull;
+import lombok.*;
```
:bug: Hvezdickovy importy se nepouzivaji. Naimportuje se kopec balastu co neni potreba.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/movie/Movie.java:
```diff
@@ -0,0 +1,79 @@
+import java.time.LocalDate;
+import java.util.UUID;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Entity
+@EqualsAndHashCode
```
:bug: Zvazit pouziti vsech lombok anotaci.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/movie/MovieContoller.java:
```diff
@@ -0,0 +1,270 @@
+            @ApiResponse(responseCode = "400",description = "Either the value in page number or page size parametrs negative or not integers"),
+            @ApiResponse(responseCode = "404",description = "The resulting list of movies is empty , most likely because the page number is too high")
+    })
+    public ArrayResponse<MovieResponse> getMoviesPage(@RequestParam String pageNumber,@RequestParam String pageSize) {
```
:bug: Proc `String` a ne `int`, kdyz ocekavana hodnota je `int`? Pak musite vytvaret z retazce intiger.

🐛 `ArrayResponse`  takato obecna trida je tady spatne. Pak kdyz budete potrebovat v nejakem endpointu vracet v nic neco specialniho, tak to pak budete mit v kazdem endpointu, kde ji mate. Zkuste neco jako:


```
class MoviesResponse {
    List<MovieResponse> movies;
}
```

jankoscak reviewed (changes requested):
Nektere pripominky jsem napsal na jednom miste, ale vyskytuji se na vicero mistech. Opravte to prosim vsude. 

Kdyz ty zmeny zapracujete, ja se opet na to podivam. Mezitym muzete zacit pracovat na rozsireni funkcionality, jak jsme se bavili. To dejte do samostatneho pull requestu pak.

Kleamens committed:
feat: Reading .csv through getResource + loggin

Instead of using absoule path is now used getResource to get the .csv file
Instead of relying on System.out for info on the program now is used Slf4j

Kleamens committed:
feat:  added .csv column naming convention

Added constants for column representation for more intuitive usage

Kleamens committed:
fix: removed excessive anotations

Removed anotations which functionality is not used

Kleamens committed:
fix: removed excessive imports

Replaced excessive imports + Replaced general imports with atomic imports

Kleamens committed:
fix: loggin + exception handling + status codes

Kleamens committed:
fix: exception handlers

 Fixed the exception handlers throwing constant 500

Kleamens committed:
fix: delegation of responisibilities + change of port in the test req

Moved the service logic from controllers to service layer + changed ports in the test requests

Kleamens committed:
refactor:  Further refactoring of controllers

Added Specific Classes for the purposes of array of responses,

Kleamens committed:
refactor: Further refactoring

Removed extraneous imports + added exception handler for invalid body input

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,156 @@
+@Service
+public class dataDownloader {
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
```
:art: Je mozne pouzit `@Slf4j` anotaci nad tridu na vygenerovani loggru.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,156 @@
+    public dataDownloader(MovieService movieService, DirectorService directorService) {
+        this.directorService = directorService;
+        this.movieService = movieService;
+    }
```
🐛 `@RequiredArgsConstructor`

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,156 @@
+
+
+@Service
+public class dataDownloader {
```
:art: `dataDownloader` -> `DataDownloader`

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/data/dataDownloader.java:
```diff
@@ -0,0 +1,156 @@
+        }catch (AssertionError e){
+            logger.error("CSV file was not found");
+            return;
+        }
```
🐛 Aby assert se vyhodnotil tak je ho potreba povolit pri spusteni. Co se teda bezne nedela. Vyjimkam treba predchazet a ne je vyvolavat. Je to draha operace. Lepsi a citelnejsi bude, kdyz udelate tohle:

```
if (resourceRelativePath == null) {
    log.error("CSV file was not found on path: {}", resourceRelativePath);
    return;
}
```

Kleamens committed:
refactor: wider use of global exception handling

Moved most of the exception handling to GlobalerroHAndler

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/utils/exceptions/GlobalErrorHandler.java:
```diff
@@ -0,0 +1,30 @@
+public class GlobalErrorHandler {
+//    For the needs of global error handling
+    @ExceptionHandler({BadInputException.class,MethodArgumentTypeMismatchException.class})
+    public ResponseEntity<?> BadInputExceptionDirector(RuntimeException ex) {
+        return new ResponseEntity<>(ex.getMessage(),BAD_REQUEST);
+
+    }
+    @ExceptionHandler(AlreadyExistsException.class)
+    public ResponseEntity exceptionDirectorExists(AlreadyExistsException ex) {
+        return new ResponseEntity<>(ex.getMessage(),CONFLICT);
+    }
+    @ExceptionHandler(NotFoundException.class)
+    public ResponseEntity exceptionDirectorNotFound(NotFoundException ex) {
+        return new ResponseEntity<>(ex.getMessage(),NOT_FOUND);
+    }
```
🐛  V `ResponseEntity` by ako body mel byt `ErrorResponse`, kterej ste si vytvořil.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/utils/exceptions/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+@Getter
+public class ErrorResponse {
+    private final String message;
+    private final HttpStatus httpStatus;
```
:bug: `HttpStatus` je v response nadbytecne. Spis by tady mel byt code pro well known error. Napríklad. kdyz budete mit request na rezervaci parkovani, ale parkovište je uz plne, tak vratite `HttpStatus` 400 (BAD REQUEST) a v `ErrorResponse` v code date `SOLD_OUT` ako code. Pak kdyz client dostane 400 `HttpStatus` s code `SOLD_OUT`, tak vi ze nema opakovat pozadavek s danyma parametrama.

jankoscak commented on moviesAPI/src/main/java/cz/mendelu/ea/xzirchuk/project/moviesAPI/domain/movie/MovieService.java:
```diff
@@ -0,0 +1,162 @@
+        }).toList());
+        return filtered_movies;
+    }
+    public List<Movie> getMoviesByDirectorName(String name){
+        List<Movie> movies = new ArrayList<>();
+
+        movieRepository.findAll().forEach(movies::add);
+        movies = movies.stream().filter(
+                movie -> movie.getDirector().getName().equals(name)
+        ).toList();
+        return  movies;
+    }
```
:bug: Tohle se da napsat na jeden radek.

```
public List<Movie> getMoviesByDirectorName(String name){
    return ...
}
```

Kleamens committed:
fix:changes to logging

now using Slf4j anotation istead of manualy creating logger inctance

Kleamens committed:
refactor: Required args annotation

Instead of writing out the constructor manualy for DI , it's done via RequiredArgsConstructor annotation

Kleamens committed:
refactor: renamed DataDownloader

changed the name of the class to follow naming convention

Kleamens committed:
refactor: Simplification of the resource path presence

Changed the use of assert in the check if the the resource exists

Kleamens committed:
fix: Using ErrorResponse entity for exception handle

Kleamens committed:
refactor: Well known errors in ErrorResponse

Instead of keeping StatusCode in ErrorResponse, the information about well known error is kept

Kleamens committed:
refactor: Simplification in getMoviesByDirectorName + CodeEnum

Code Enum still returns the same values

Kleamens committed:
refactor: removed excessive imports

jankoscak commented:
I kdyz jeste mel nejake pripominky, tak se posunem dal. Mergnu to.
